src/libmcrx.c

```diff
@@ -1055,3 +1075,387 @@ mcrx_is_system_error(enum mcrx_error_code err) {
   return err >= MCRX_ERR_SYSCALL_BIND;
 }
 
+/**
+ * mcrx_mnatmap_new:
+ * @mnatmapp: pointer to be filled with the new mnat context
+ *
+ * Create a new mnat context.
+ *
+ * Returns: error code
+ **/
+MCRX_EXPORT enum mcrx_error_code mcrx_mnatmap_new(
+    struct mcrx_mnatmap **mnatmapp) {
+  if (mnatmapp == NULL) {
+    return MCRX_ERR_NULLARG;
+  }
+
+  struct mcrx_mnatmap *mnat_ctx;
+  mnat_ctx = calloc(1, sizeof(struct mcrx_mnatmap));
+  if (!mnat_ctx) {
+    return MCRX_ERR_NOMEM;
+  }
+
+  mnat_ctx->refcount = 1;
+  LIST_INIT(&mnat_ctx->mnats_head);
+
+  *mnatmapp = mnat_ctx;
+  return MCRX_ERR_OK;
+}
+
+/**
+ * mcrx_mnatmap_apply:
+ * Apply the mnat context to mcrx context
+ *
+ **/
+MCRX_EXPORT enum mcrx_error_code mcrx_mnatmap_apply(struct mcrx_ctx *ctx,
+    struct mcrx_mnatmap *mnatmap) {
+  if (ctx == NULL) {
+    return MCRX_ERR_NULLARG;
+  }
+  if (ctx->mnat_map == NULL) {
+    // no mnat running time mcrx context yet
+    mcrx_mnatmap_clone(mnatmap, &ctx->mnat_map);
+    struct mcrx_subscription *sub;
+    LIST_FOREACH(sub, &ctx->subs_head, sub_entries)
+    {
+      // find new entry based on the new mnat context
+      struct mcrx_mnat_entry *new_entry =
+          mcrx_mnatmap_find_entry_from_subscription(sub, ctx->mnat_map);
+      if (sub->joined) {
+        if (new_entry == NULL
+            || !mcrx_mnatmap_entry_local_equal(sub->mnat_entry, new_entry)) {
+          mcrx_subscription_leave(sub);
+          sub->mnat_entry = new_entry;
+          if (sub->mnat_entry
+              && !mcrx_mnatmap_entry_unresolved(sub->mnat_entry)) {
+            // rejoin with new entry
+            info(sub->ctx, "subscription %p rejoin new mnat entry %p\n", (void *)sub, (void*)sub->mnat_entry);
+            enum mcrx_error_code err = mcrx_subscription_native_join(sub);
+            if (err == MCRX_ERR_OK) {
+              ctx->live_subs++;
+            }
+          }
+        }
+      }
+    }
+  } else {
+    struct mcrx_mnatmap *clone_context;
+    mcrx_mnatmap_clone(mnatmap, &clone_context);
+    struct mcrx_subscription *sub;
+    LIST_FOREACH(sub, &ctx->subs_head, sub_entries) {
+      // find new entry based on the new mnat context
+      struct mcrx_mnat_entry *new_entry = mcrx_mnatmap_find_entry_from_subscription(sub, clone_context);
+      if (sub->joined) {
+        if (new_entry == NULL || !mcrx_mnatmap_entry_local_equal(sub->mnat_entry, new_entry)) {
+          mcrx_subscription_leave(sub);
+          sub->mnat_entry = new_entry;
+          if (sub->mnat_entry
+              && !mcrx_mnatmap_entry_unresolved(sub->mnat_entry)) {
+            // rejoin with new entry
+            info(sub->ctx, "subscription %p rejoin new mnat entry %p\n", (void *)sub, (void*)sub->mnat_entry);
+            enum mcrx_error_code err = mcrx_subscription_native_join(sub);
```

Comment:
As discussed, we need to add a callback to the subscription to report these native join errors, and we need a state for "global join requested but not resolved".  Since this is complicated, we'll discuss more Thursday...